File: coursera exercises/dijkstra.py
#Uses python3

import sys
import queue
from heapq import heapify, heappop, heappush

class priority_queue():
    def __init__(self):
        self.queue = list()
        heapify(self.queue)
        self.index = dict()
    def push(self, priority, label):
        if label in self.index:
            self.queue = [(w,l)
                for w,l in self.queue if l!=label]
            heapify(self.queue)
        heappush(self.queue, (priority, label))
        self.index[label] = priority
    def pop(self):
        if self.queue:
            return heappop(self.queue)
    def __contains__(self, label):
        return label in self.index
    def __len__(self):
        return len(self.queue)

def distance(adj, cost, s, t):
    graphsize = len(adj)
    adj_cost = [dict(zip(adj[i],cost[i])) for i in range(graphsize)]
    inf = float('inf')
    known = set()
    priority = priority_queue()
    path = {s:s}
    
    for vertex in range(graphsize):
        if vertex == s:
            priority.push(0,vertex)
        else:
            priority.push(inf,vertex)
    last_node = s
    
    while last_node != t:
        (weight, actual_node) = priority.pop()
        if actual_node not in known:
            for next_node in adj_cost[actual_node]:
                upto_actual = priority.index[actual_node]
                upto_next = priority.index[next_node]
                to_next = upto_actual + \
                adj_cost[actual_node][next_node]
                if to_next < upto_next:
                    priority.push(to_next, next_node)
                    path[next_node] = actual_node
            last_node = actual_node
            known.add(actual_node)
    return priority.index[t] if priority.index[t] != inf else -1


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s, t = data[0] - 1, data[1] - 1
    print(distance(adj, cost, s, t))

File: coursera exercises/Dijkstra.java
import java.util.*;

class PriorityQueueWithDecrease {
    private PriorityQueue<int[]> pq;
    private Map<Integer, Integer> index;

    public PriorityQueueWithDecrease(int n) {
        pq = new PriorityQueue<>(n, Comparator.comparingInt(a -> a[1]));
        index = new HashMap<>();
    }

    public void push(int vertex, int priority) {
        if (index.containsKey(vertex)) {
            pq.remove(new int[]{vertex, index.get(vertex)});
        }
        pq.add(new int[]{vertex, priority});
        index.put(vertex, priority);
    }

    public int[] pop() {
        return pq.poll();
    }

    public boolean contains(int vertex) {
        return index.containsKey(vertex);
    }

    public int size() {
        return pq.size();
    }
}

public class Dijkstra {
    private static long distance(ArrayList<Integer>[] adj, ArrayList<Integer>[] cost, int s, int t) {
        int graphSize = adj.length;
        Map<Integer, Integer>[] adjCost = new Map[graphSize];
        for (int i = 0; i < graphSize; i++) {
            adjCost[i] = new HashMap<>();
            for (int j = 0; j < adj[i].size(); j++) {
                adjCost[i].put(adj[i].get(j), cost[i].get(j));
            }
        }

        long[] distance = new long[graphSize];
        Arrays.fill(distance, Long.MAX_VALUE);
        distance[s] = 0;

        int[] path = new int[graphSize];
        Arrays.fill(path, -1);
        path[s] = s;

        Set<Integer> known = new HashSet<>();
        PriorityQueueWithDecrease pq = new PriorityQueueWithDecrease(graphSize);
        pq.push(s, 0);

        while (pq.size() > 0) {
            int[] top = pq.pop();
            int actualNode = top[0];
            long weight = top[1];

            if (known.contains(actualNode)) {
                continue;
            }

            known.add(actualNode);

            for (Map.Entry<Integer, Integer> entry : adjCost[actualNode].entrySet()) {
                int nextNode = entry.getKey();
                int edgeCost = entry.getValue();
                if (distance[actualNode] + edgeCost < distance[nextNode]) {
                    distance[nextNode] = distance[actualNode] + edgeCost;
                    pq.push(nextNode, (int) distance[nextNode]);
                    path[nextNode] = actualNode;
                }
            }
        }

        return distance[t] != Long.MAX_VALUE ? distance[t] : -1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        ArrayList<Integer>[] cost = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
            cost[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y, w;
            x = scanner.nextInt();
            y = scanner.nextInt();
            w = scanner.nextInt();
            adj[x - 1].add(y - 1);
            cost[x - 1].add(w);
        }
        int x = scanner.nextInt() - 1;
        int y = scanner.nextInt() - 1;
        System.out.println(distance(adj, cost, x, y));
    }
}

